### PR TITLE
[V1][Scheduler] Clamp KV connector match to the prompt suffix

### PR DESCRIPTION
**Pull request by n-dlms**
## Summary

`Scheduler.schedule()` can crash with a bare `AssertionError` when a KV connector is attached and prefix caching produced a local hit:

- `assert num_new_tokens > 0` (scheduler.py ~914) when the connector's match plus the local hit reaches the end of the prompt (zero local tokens left for the prefill step), or
- `assert num_computed_tokens <= request.num_tokens` (scheduler.py ~832) when the match plus the local hit *exceeds* the prompt length.

The local prefix lookup already caps hits at `num_tokens - 1` (`get_computed_blocks`, kv_cache_manager.py:259) because the last token must always be recomputed to obtain logits — and the async-load promotion path applies the same rule ("on a full prompt hit, we need to re-compute the last token", scheduler.py ~2671). The connector-reported external match was the only unbounded input: it is added to the local hit without being clamped to the remaining prompt suffix.

This is reachable through vLLM's own test connector: `MockKVConnector` returns a fixed `matched_tokens` regardless of the local hit (tests/v1/kv_connector/unit/utils.py), and any connector match of `M` tokens on top of a local hit `L` trips the asserts whenever `L + M >= num_tokens`. The same code exists in upstream `main` (both asserts, no clamp).

## Fix

In the prefill branch of `Scheduler.schedule()`, clamp the connector match to the prompt suffix that is not locally cached, mirroring the `num_tokens - 1` cap of the local lookup:

```python
num_external_computed_tokens = min(
    num_external_computed_tokens,
    max(request.num_tokens - num_new_local_computed_tokens - 1, 0),
)
if num_external_computed_tokens == 0:
    load_kv_async = False
```

- At least one token of local prefill remains, so `num_new_tokens > 0` holds.
- `num_computed_tokens <= num_tokens` holds; the `num_tokens - 1` bound matches the documented "recompute the last token" rule.
- When the clamp leaves nothing to load, `load_kv_async` is cleared so the async-load assert (`num_external_computed_tokens > 0`) stays satisfiable.
- `connector_prefix_cache_hits` (stats) now records the clamped value actually used.

No behavior change when the connector reports a legal match: the clamp only reduces matches that would overrun the prompt.

## Tests

Two regression tests in `tests/v1/core/test_scheduler.py` cover both failure faces (each fails with `AssertionError` before the fix, passes after):

- `test_connector_match_clamped_to_prompt_suffix` — a match that reaches the end of the prompt (zero local tokens) is clamped to the remaining suffix minus one; the request then decodes normally.
- `test_connector_match_exceeding_prompt_clamped` — a block-aligned local hit plus a longer match would exceed the prompt; the match is clamped and the request finishes prefill.

Commands and results:

```
# regression tests fail without the fix (stashed scheduler.py):
.venv/bin/python -m pytest tests/v1/core/test_scheduler.py::test_connector_match_clamped_to_prompt_suffix tests/v1/core/test_scheduler.py::test_connector_match_exceeding_prompt_clamped -q
# -> 2 failed (AssertionError inside Scheduler.schedule())

# with the fix:
# -> 2 passed

# full scheduler test file (incl. 66 connector/kv/prefix-cache/throttle tests):
.venv/bin/python -m pytest tests/v1/core/test_scheduler.py -q -p no:cacheprovider
# -> 139 passed, 16 warnings in 229.93s
```

No model-affecting change: this only bounds token accounting at scheduling time; sampling/output paths are untouched, so no model eval is needed.

## AI assistance

This change was developed with AI assistance (opencode). The submitting author reviewed every line and ran the tests above.

Not a duplicate: searched open PRs/issues for connector match clamping / `get_num_new_matched_tokens` bounds / the two asserts — nothing open addresses this.

### CODE REVIEW

**Review by njhill — CHANGES_REQUESTED** [review]
I don't think this is the correct fix.

This condition also won't actually occur with any of the real connectors.

**Comment by njhill** [comment]
@n-dlms perhaps you could add a clamp within `MockKVConnector`  instead?

**Comment by n-dlms** [comment]
Closing this in favor of #55113, and agreeing with the review that the connector is the right layer for this fix.

Since the review, I traced the crash end to end with `MooncakeStoreConnector` in `load_async=False` mode: when the lookup reports the full prompt length, `Scheduler.schedule` trips `assert num_new_tokens > 0` at `scheduler.py:923`. I also confirmed that clamping in the scheduler alone is not sufficient there, because MooncakeStore caches the raw lookup value in `load_specs` and asserts round trip equality in `update_state_after_alloc`, so the fix has to live where the value is produced and cached.

#55113 applies the `num_tokens - 1` cap inside `MooncakeStoreScheduler.get_num_new_matched_tokens`, the same idiom the example, HF3FS, LMCache, MoRIIO READ and decode bench connectors already use, with two regression tests that fail without the fix and pass with it. Verified locally with 91 related tests green and zero new failures against main, plus ruff clean. No scheduler code is touched.

Thanks for the layering call. It was the right one.
